xen/arch/x86/tpm.c

```diff
@@ -383,6 +397,338 @@ static void *create_log_event12(struct txt_ev_log_container_12 *evt_log,
+            p += SHA1_DIGEST_SIZE;
+        } else if ( hash_type == TPM_ALG_SHA256 ) {
+            if ( sha256_digest != NULL)
+                memcpy(sha256_digest, p, SHA256_DIGEST_SIZE);
+            p += SHA256_DIGEST_SIZE;
+        } else {
+            break;
+        }
+    }
+
+    rc = 0;
+
+error:
+    relinquish_locality(loc);
+    return rc;
+}
+
+#else
+
+union tpm2_cmd_rsp {
+    /* Enough for a couple of hashes. */
+    uint8_t b[sizeof(struct tpm2_extend_cmd) + 64];
+    struct tpm_cmd_hdr c;
+    struct tpm_rsp_hdr r;
+    struct tpm2_extend_cmd extend_c;
+    struct tpm2_extend_rsp extend_r;
+};
+
+static uint32_t tpm20_pcr_extend(unsigned loc, uint8_t *buf, unsigned size,
+                                 uint32_t pcr_handle,
+                                 const uint8_t *sha1_digest,
+                                 const uint8_t *sha256_digest)
+{
+    union tpm2_cmd_rsp cmd_rsp;
+    unsigned o_size;
+    unsigned hash_count;
+    uint8_t *p;
+
+    cmd_rsp.extend_c = (struct tpm2_extend_cmd) {
+        .h.tag = swap16(TPM_ST_SESSIONS),
+        .h.ordinal = swap32(TPM2_PCR_Extend),
+        .pcrHandle = swap32(pcr_handle),
+        .sessionHdrSize = swap32(sizeof(struct tpm2_session_header)),
+        .pcrSession.handle = swap32(TPM_RS_PW),
+    };
+
+    hash_count = 0;
+    p = cmd_rsp.extend_c.hashes;
+
+    if ( sha1_digest != NULL ) {
+        ++hash_count;
+        PUT_16BIT(p, 512);
+        PUT_BYTES(p, sha1_digest, SHA1_DIGEST_SIZE);
+    }
+    if ( sha256_digest != NULL ) {
+        ++hash_count;
+        PUT_16BIT(p, TPM_ALG_SHA256);
+        PUT_BYTES(p, sha256_digest, SHA256_DIGEST_SIZE);
+    }
+
+    cmd_rsp.extend_c.hashCount = swap32(hash_count);
+
+    /* Fill in command size (size of the whole buffer). */
+    cmd_rsp.extend_c.h.paramSize = swap32(sizeof(cmd_rsp.extend_c) +
+                                          (p - cmd_rsp.extend_c.hashes)),
+
+    o_size = sizeof(cmd_rsp);
+    printk("o_size = %d\n", o_size);
+    send_cmd(loc, cmd_rsp.b, swap32(cmd_rsp.c.paramSize), &o_size);
+    printk("o_size = %d\n", o_size);
+
+    return swap32(cmd_rsp.r.returnCode);
+}
+
+static uint32_t tpm_hash_extend20(unsigned loc, uint8_t *buf, unsigned size,
+                                  unsigned pcr, uint8_t *sha1_digest,
+                                  uint8_t *sha256_digest)
+{
+    static int has_sha1 = -1;
+
+    uint32_t rc;
+
+    request_locality(loc);
+
+    if ( has_sha1 < 0 ) {
+        uint8_t test_sha1_digest[SHA1_DIGEST_SIZE] = {0};
+        /* This is a valid way of checking hash support, using it to not
+         * implement TPM2_GetCapability(). */
+        rc = tpm20_pcr_extend(loc, buf, /*size=*/0, /*pcr_handle=*/TPM_RH_NULL,
+                              test_sha1_digest, /*sha256_digest=*/NULL);
+        has_sha1 = (rc == 0);
+    }
+
+    if (has_sha1)
+        sha1_hash(buf, size, (uint32_t *)sha1_digest);
+    else
+        sha1_digest = NULL;
```

Comment:
This will have to be a bit more complicated for event logging. I know this will be a separate PR, but I'll leave the comment here. You don't have to fix it here, but this is something to keep in mind.

Event log isn't initialized by Xen, it has to follow whatever format already exists. The header has information about algorithms contained inside, so information about SHA1 availability should come from outside of this function, assuming the header reflects what TPM implements, of course. I don't know how ACM processes it, in particular if it creates SHA1 entries in log if it isn't available on TPM (apparently there are already PTTs that allow to turn SHA1 off). So there are 4 possibilities:
- both TPM and event log support SHA1 -> no problem, hash it.
- both TPM and event log don't support SHA1 -> no problem, don't hash it.
- event log supports, TPM doesn't -> we probably need to write something to the log, either all 00s, FFs or some distinguishable magic value so anyone reading the log will notice that something is not right. We can't trust that event log is already zeroed, that would require trusting pre-DL event code.
- TPM supports, event log doesn't -> we can extend the PCR but not log it, or skip extending SHA1. In any case, PCR value can't be replayed, and we can't really check whether ACM extended this bank or not, since TPM will extend all banks for PCR-17 with DCE's hash anyway. I don't see any value with extending SHA1 in this case, but maybe @dpsmith can think of something.

Reply:
> event log supports, TPM doesn't -> we probably need to write something to the log, either all 00s, FFs or some distinguishable magic value so anyone reading the log will notice that something is not right. We can't trust that event log is already zeroed, that would require trusting pre-DL event code.

TXT spec has this:

```
Unsupported bank capping value

OneDigest is extended into PCR bank using an algorithm that is not supported by SINIT SW.
```

```
OneDigest is sequence of bytes: 01,00,00…00 with a total size equal to HashAlgId_DIGEST_SIZE
of capped TPM bank
```

Reply:
We can use this value, although this is the opposite situation.